src/Graph.elm

```diff
@@ -960,34 +961,42 @@ guidedDfs :
     -> ( acc, Graph n e )
 guidedDfs selectNeighbors visitNode startingSeeds startingAcc startingGraph =
     let
-        go seeds acc graph =
+        go : List NodeId -> Set NodeId -> acc -> Graph n e -> ( acc, Set NodeId, Graph n e )
+        go seeds visited acc graph =
             case seeds of
                 [] ->
                     -- We are done with this connected component, so we return acc and the rest of the graph
-                    ( acc, graph )
+                    ( acc, visited, graph )
 
                 next :: seeds1 ->
-                    case get next graph of
-                        -- This can actually happen since we don't filter for already visited nodes.
-                        -- That would be an opportunity for time-memory-tradeoff.
-                        -- E.g. Passing along a set of visited nodeIds.
-                        Nothing ->
-                            go seeds1 acc graph
+                    if Set.member next visited then
+                        go seeds1 visited acc graph
 
-                        Just ctx ->
-                            let
-                                ( accAfterDiscovery, finishNode ) =
-                                    visitNode ctx acc
-
-                                ( accBeforeFinish, graph1 ) =
-                                    go (selectNeighbors ctx) accAfterDiscovery (remove next graph)
-
-                                accAfterFinish =
-                                    finishNode accBeforeFinish
-                            in
-                            go seeds1 accAfterFinish graph1
+                    else
+                        case get next graph of
+                            -- This can actually happen since we don't filter for already visited nodes.
+                            -- That would be an opportunity for time-memory-tradeoff.
+                            -- E.g. Passing along a set of visited nodeIds.
+                            Nothing ->
+                                go seeds1 (Set.insert next visited) acc graph
```

Comment:
I don't think this comment is accurate any longer. This code path is only taken when we try to access a node that is not part of the graph.